File: Source/Client/Modules/Systems/Party.vb
Imports System.Drawing
Imports Mirage.Sharp.Asfw
Imports Core

Module Party

#Region "Database"

    Sub ClearParty()
        Type.Party = New PartyStruct With {
            .Leader = 0,
            .MemberCount = 0
        }
        ReDim Type.Party.Member(MAX_PARTY_MEMBERS)
    End Sub

#End Region

#Region "Incoming Packets"

    Sub Packet_PartyInvite(ByRef data() As Byte)
        Dim name As String
        Dim buffer As New ByteStream(data)

        name = buffer.ReadString
        Dialogue("Party Invite", name & " has invited you to a party.", "Would you like to join?", DialogueType.Party, DialogueStyle.YesNo)

        buffer.Dispose()
    End Sub

    Sub Packet_PartyUpdate(ByRef data() As Byte)
        Dim i As Integer, inParty As Integer
        Dim buffer As New ByteStream(data)

        inParty = buffer.ReadInt32

        ' exit out if we're not in a party
        If inParty = 0 Then
            ClearParty()
            Gui.UpdatePartyInterface
            ' exit out early
            buffer.Dispose()
            Exit Sub
        End If

        ' carry on otherwise
        Type.Party.Leader = buffer.ReadInt32
        For i = 1 To MAX_PARTY_MEMBERS
            Type.Party.Member(I) = buffer.ReadInt32
        Next
        Type.Party.MemberCount = buffer.ReadInt32

        Gui.UpdatePartyInterface

        buffer.Dispose()
    End Sub

    Sub Packet_PartyVitals(ByRef data() As Byte)
        Dim playerNum As Integer, partyindex As Integer
        Dim buffer As New ByteStream(data)

        ' which player?
        playerNum = buffer.ReadInt32

        ' find the party number
        For i = 1 To MAX_PARTY_MEMBERS
            If Type.Party.Member(I) = playerNum Then
                partyindex = I
            End If
        Next

        ' exit out if wrong data
        If partyindex <= 0 Or partyindex > MAX_PARTY_MEMBERS Then Exit Sub

        ' set vitals
        For i = 1 To VitalType.Count - 1
            Type.Player(playerNum).Vital(i) = buffer.ReadInt32
        Next

        UpdatePartyBars

        buffer.Dispose()
    End Sub

#End Region

#Region "Outgoing Packets"

    Friend Sub SendPartyRequest(name As String)
        Dim buffer As New ByteStream(4)
        buffer.WriteInt32(ClientPackets.CRequestParty)
        buffer.WriteString((name))

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

    Friend Sub SendAcceptParty()
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CAcceptParty)

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

    Friend Sub SendDeclineParty()
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CDeclineParty)

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

    Friend Sub SendLeaveParty()
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CLeaveParty)

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

    Friend Sub SendPartyChatMsg(text As String)
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CPartyChatMsg)
        buffer.WriteString((text))

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

#End Region

End Module

File: Source/Client/Game/Systems/Party.cs
﻿using Core;
using Mirage.Sharp.Asfw;

namespace Client
{

    public class Party
    {

        #region Database

        public static void ClearParty()
        {
            Data.MyParty = new Core.Type.Party()
            {
                Leader = 0,
                MemberCount = 0
            };
            Data.MyParty.Member = new int[5];
        }

        #endregion

        #region Incoming Packets

        public static void Packet_PartyInvite(ref byte[] data)
        {
            string name;
            var buffer = new ByteStream(data);

            name = buffer.ReadString();
            GameLogic.Dialogue("Party Invite", name + " has invited you to a party.", "Would you like to join?", (byte)DialogueType.PartyInvite, (byte)DialogueStyle.YesNo);

            buffer.Dispose();
        }

        public static void Packet_PartyUpdate(ref byte[] data)
        {
            int i;
            int inParty;
            var buffer = new ByteStream(data);

            inParty = buffer.ReadInt32();

            // exit out if we're not in a party
            if (inParty == -1)
            {
                ClearParty();
                Gui.UpdatePartyInterface();
                // exit out early
                buffer.Dispose();
                return;
            }

            // carry on otherwise
            Data.MyParty.Leader = buffer.ReadInt32();
            for (i = 0; i < Constant.MAX_PARTY_MEMBERS; i++)
                Data.MyParty.Member[i] = buffer.ReadInt32();
            Data.MyParty.MemberCount = buffer.ReadInt32();

            Gui.UpdatePartyInterface();

            buffer.Dispose();
        }

        public static void Packet_PartyVitals(ref byte[] data)
        {
            int playerNum;
            var partyindex = -1;
            var buffer = new ByteStream(data);

            // which player?
            playerNum = buffer.ReadInt32();

            // find the party number
            for (int i = 0; i < Constant.MAX_PARTY_MEMBERS; i++)
            {
                if (Data.MyParty.Member[i] == playerNum)
                {
                    partyindex = i;
                }
            }

            // exit out if wrong data
            if (partyindex < 0 | partyindex >= Constant.MAX_PARTY_MEMBERS)
                return;

            // set vitals
            var vitalCount = Enum.GetNames(typeof(Core.Vital)).Length;
            for (int i = 0; i < vitalCount; i++)
                Core.Data.Player[playerNum].Vital[i] = buffer.ReadInt32();

            GameLogic.UpdatePartyBars();

            buffer.Dispose();
        }

        #endregion

        #region Outgoing Packets

        public static void SendPartyRequest(string name)
        {
            var buffer = new ByteStream(4);
            buffer.WriteInt32((int)Packets.ClientPackets.CRequestParty);
            buffer.WriteString(name);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        public static void SendAcceptParty()
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CAcceptParty);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        public static void SendDeclineParty()
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CDeclineParty);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        public static void SendLeaveParty()
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CLeaveParty);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        public static void SendPartyChatMsg(string text)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CPartyChatMsg);
            buffer.WriteString(text);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        #endregion

    }
}